File: Algorithms/some_intern_contests/tinkoff_winter/task_6/brute_force.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <set>

using namespace std;

typedef pair<int, int> Point;

bool is_non_degenerate(const Point& p1, const Point& p2, const Point& p3) {
    return (p2.first - p1.first) * (p3.second - p1.second) != 
           (p3.first - p1.first) * (p2.second - p1.second);
}

int backtrack(int index, set<int>& used_points, int count, 
              const vector<vector<int>>& valid_triangles) {
    if (index == valid_triangles.size()) {
        return count;
    }
    
    // Try including the current triangle
    const auto& current_triangle = valid_triangles[index];
    bool can_include = true;
    for (int point : current_triangle) {
        if (used_points.find(point) != used_points.end()) {
            can_include = false;
            break;
        }
    }
    
    int include = 0;
    if (can_include) {
        set<int> new_used_points = used_points;
        for (int point : current_triangle) {
            new_used_points.insert(point);
        }
        include = backtrack(index + 1, new_used_points, count + 1, valid_triangles);
    }
    
    // Try excluding the current triangle
    int exclude = backtrack(index + 1, used_points, count, valid_triangles);
    
    return max(include, exclude);
}

int solve(int n, const vector<Point>& points) {
    vector<vector<int>> all_triangles;
    for (int i = 0; i < n - 2; ++i) {
        for (int j = i + 1; j < n - 1; ++j) {
            for (int k = j + 1; k < n; ++k) {
                all_triangles.push_back({i, j, k});
            }
        }
    }
    
    
    vector<vector<int>> valid_triangles;
    for (const auto& triangle : all_triangles) {
        if (is_non_degenerate(points[triangle[0]], points[triangle[1]], points[triangle[2]])) {
            valid_triangles.push_back(triangle);
        }
    }
    
    set<int> used_points;
    return backtrack(0, used_points, 0, valid_triangles);
}

int main() {
    int n;
    cin >> n;
    
    vector<Point> points(n);
    for (int i = 0; i < n; ++i) {
        cin >> points[i].first >> points[i].second;
    }
    
    cout << solve(n, points) << endl;
    
    return 0;
}

File: Algorithms/some_intern_contests/tinkoff_winter/task_6/brute_force.py
from itertools import combinations

def is_non_degenerate(p1, p2, p3):
    return (p2[0] - p1[0]) * (p3[1] - p1[1]) != (p3[0] - p1[0]) * (p2[1] - p1[1])

def solve(n, points):
    valid_triangles = []
    for i in range(n):
        for j in range(i+1, n):
            for k in range(j+1, n):
                if is_non_degenerate(points[i], points[j], points[k]):
                    valid_triangles.append((i, j, k))
    
    # Sort triangles by some heuristic (e.g., perimeter)
    def triangle_perimeter(t):
        return sum(abs(points[t[i]][0] - points[t[j]][0]) + abs(points[t[i]][1] - points[t[j]][1]) 
                   for i, j in [(0,1), (1,2), (2,0)])
    
    valid_triangles.sort(key=triangle_perimeter, reverse=True)
    
    def backtrack(index, used_mask, count):
        if count * 3 + n - bin(used_mask).count('1') < count * 3:
            return count
        
        if index == len(valid_triangles):
            return count
        
        current_triangle = valid_triangles[index]
        triangle_mask = (1 << current_triangle[0]) | (1 << current_triangle[1]) | (1 << current_triangle[2])
        
        if used_mask & triangle_mask == 0:
            include = backtrack(index + 1, used_mask | triangle_mask, count + 1)
        else:
            include = 0
        
        exclude = backtrack(index + 1, used_mask, count)
        
        return max(include, exclude)
    
    return backtrack(0, 0, 0)

def main():
    n = int(input())
    points = [tuple(map(int, input().split())) for _ in range(n)]
    print(solve(n, points))

if __name__ == "__main__":
    main()